copycat/runtime/src/main/java/org/apache/kafka/copycat/runtime/distributed/KafkaConfigStorage.java

```diff
@@ -0,0 +1,438 @@
+import org.apache.kafka.copycat.data.Schema;
+import org.apache.kafka.copycat.data.SchemaAndValue;
+import org.apache.kafka.copycat.data.SchemaBuilder;
+import org.apache.kafka.copycat.data.Struct;
+import org.apache.kafka.copycat.errors.CopycatException;
+import org.apache.kafka.copycat.errors.DataException;
+import org.apache.kafka.copycat.storage.Converter;
+import org.apache.kafka.copycat.util.Callback;
+import org.apache.kafka.copycat.util.ConnectorTaskId;
+import org.apache.kafka.copycat.util.KafkaBasedLog;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * <p>
+ *     Provides persistent storage of Copycat connector configurations in a Kafka topic.
+ * </p>
+ * <p>
+ *     This class manages both connector and task configurations. It tracks three types of configuration entries:
+ *
+ *     1. Root config: list of connector IDs, each with the current number of tasks (Kafka key: root).
+ *                          This config is used only to track which task configs are in a consistent state and is
+ *                          therefore effectively ephemeral -- if it were deleted, we would treat everything as
+ *                          inconsistent until all connector's tasks were updated and a new record specifying
+ *                          the number of tasks per connector was written.
+ *     2. Connector config: map of string -> string configurations passed to the Connector class, with support for
+ *                          expanding this format if necessary. (Kafka key: connector-[connector-id]).
+ *                          These configs are *not* ephemeral. They represent the source of truth. If the entire Copycat
+ *                          cluster goes down, this is all that is really needed to recover.
+ *     3. Task configs: map of string -> string configurations passed to the Task class, with support for expanding
+ *                          this format if necessary. (Kafka key: task-[connector-id]-[task-id]).
+ *                          These configs are ephemeral; they are stored here to a) disseminate them to all workers while
+ *                          ensuring agreement and b) to allow faster cluster/worker recovery since the common case
+ *                          of recovery (restoring a connector) will simply result in the same configuration as before
+ *                          the failure.
+ * </p>
+ * <p>
+ *     The root configuration is used to trigger *task* updates since we require a combination of a compacted topic and atomic
+ *     updates. To accomplish this, we do *not* apply updates to each configuration as we read them off the log. Instead,
```

Comment:
By "we", I meant this class. I've rewritten those couple of sentences a bit to try to make that clear.

This would be a lot simpler if we had txn support :)